<a href="https://colab.research.google.com/github/lee-jee-hyeon/lee-jee-hyeon/blob/main/python_numerical_computing_with_numpy_ipynb%EC%9D%98_%EC%82%AC%EB%B3%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Working in Colab

In [ ]:
from google.colab import drive
drive.mount("/content/gdrive")

Mounted at /content/gdrive


In [ ]:
root_path = '/content/gdrive/My Drive/ML/Basic'

## 임가현 멘토 질문 : 데이터분석가가 Numpy를 사용하는 이유?

# Python 및 Numpy를 사용한 수치 계산

![](https://i.imgur.com/mg8O3kd.png)

## 숫자 데이터 작업
데이터 분석의 "데이터"는 일반적으로 주가, 판매 수치, 센서 측정, 스포츠 점수, 데이터베이스 테이블 등과 같은 수치 데이터를 나타냅니다. Numpy 라이브러리는 Python의 수치 컴퓨팅을 위한 특수 데이터 구조, 기능 및 기타 도구를 제공합니다. . 숫자 데이터 작업에 Numpy를 사용하는 이유와 방법을 알아보기 위해 예제를 살펴보겠습니다.

> 온도, 강우량 및 습도와 같은 기후 데이터를 사용하여 해당 지역이 사과 재배에 적합한지 결정한다고 가정합니다. 이를 위한 간단한 접근 방식은 사과의 연간 수확량(헥타르당 톤)과 평균 온도(화씨 단위), 강우량(밀리미터 단위) 및 평균 상대 습도(백분율 단위)와 같은 기후 조건 간의 관계를 공식화하는 것입니다. 선형 방정식으로.

> yield_of_apples = w1 * 온도 + w2 * 강우량 + w3 * 습도

우리는 사과의 수확량을 온도, 강우량 및 습도의 가중치 합으로 표현하고 있습니다. 실제 관계가 반드시 선형일 필요는 없고 다른 요인이 관련될 수 있으므로 이 방정식은 근사치입니다. 그러나 이와 같은 단순한 선형 모델은 종종 실제로 잘 작동합니다.

과거 데이터에 대한 일부 정적 분석을 기반으로 w1, w2 및 w3 가중치에 대해 합리적인 값을 제시할 수 있습니다. 다음은 값 집합의 예입니다.

In [ ]:
w1, w2, w3 = 0.3, 0.2, 0.5

지역에 대한 일부 기후 데이터가 주어지면 이제 사과 수확량을 예측할 수 있습니다. 다음은 몇 가지 샘플 데이터입니다.
<img src="https://i.imgur.com/TXPBiqv.png" style="width:360px;"><br>
시작하려면 지역의 기후 데이터를 기록하기 위해 몇 가지 변수를 정의할 수 있습니다.

In [ ]:
kanto_temp = 73
kanto_rainfall = 67
kanto_humidity = 43

이제 이러한 변수를 선형 방정식에 대입하여 사과 수확량을 예측할 수 있습니다.

In [ ]:
kanto_yield_apples = kanto_temp * w1 + kanto_rainfall * w2 + kanto_humidity * w3
kanto_yield_apples

56.8

In [ ]:
print("The expected yield of apples in Kanto region is {} tons per hectare.".format(kanto_yield_apples))

The expected yield of apples in Kanto region is 56.8 tons per hectare.


간토 지역의 예상 사과 수확량은 헥타르당 56.8톤입니다.
여러 지역에 대해 위의 계산을 약간 더 쉽게 수행하기 위해 각 지역의 기후 데이터를 벡터, 즉 숫자 목록으로 나타낼 수 있습니다.

In [ ]:
kanto = [73, 67, 43]
johto = [91, 88, 64]
hoenn = [87, 134, 58]
sinnoh = [102, 43, 37]
unova = [69, 96, 70]

각 벡터의 세 숫자는 각각 온도, 강우량 및 습도 데이터를 나타냅니다.

공식에 사용된 가중치 집합을 벡터로 나타낼 수도 있습니다.

In [ ]:
weights = [w1, w2, w3]

이제 우리는 기후 데이터와 각각의 가중치가 주어진 사과(또는 다른 작물)의 수확량을 계산하는 함수 작물 수확량을 작성할 수 있습니다.

In [ ]:
def crop_yield(region, weights):
    result = 0
    for x, w in zip(region, weights):
        result += x * w
    return result

In [ ]:
crop_yield(kanto, weights)

56.8

In [ ]:
crop_yield(johto, weights)

76.9

In [ ]:
crop_yield(unova, weights)

74.9

### Python 목록에서 Numpy 배열로 이동
Crop_yield에 의해 수행되는 계산(두 벡터의 요소별 곱셈과 결과의 합을 취함)을 내적이라고도 합니다. 여기에서 내적에 대해 자세히 알아보십시오. https://www.khanacademy.org/math/linear-algebra/vectors-and-spaces/dot-cross-products/v/vector-dot-product-and-vector-length .

Numpy 라이브러리는 두 벡터의 내적을 계산하는 내장 함수를 제공합니다. 그러나 먼저 목록을 Numpy 배열로 변환해야 합니다.

pip 패키지 관리자를 사용하여 Numpy 라이브러리를 설치해 보겠습니다.

In [ ]:
#!pip install numpy --upgrade --quiet

다음으로 numpy 모듈을 가져오겠습니다. 별칭 np로 numpy를 가져오는 것이 일반적입니다.

In [ ]:
import numpy as np

이제 np.array 함수를 사용하여 Numpy 배열을 생성할 수 있습니다.

In [ ]:
kanto = np.array([73, 67, 43])

In [ ]:
kanto

array([73, 67, 43])

In [ ]:
weights = np.array([w1, w2, w3])

In [ ]:
weights

array([0.3, 0.2, 0.5])

Numpy 배열에는 ndarray 유형이 있습니다.

In [ ]:
type(kanto)

numpy.ndarray

In [ ]:
type(weights)

numpy.ndarray

목록과 마찬가지로 Numpy 배열은 인덱싱 표기법 []을 지원합니다.

In [ ]:
weights[0]

0.3

In [ ]:
kanto[2]

43

### Numpy 배열에서 작동
이제 np.dot 함수를 사용하여 두 벡터의 내적을 계산할 수 있습니다.

In [ ]:
np.dot(kanto, weights)

56.8

Numpy 배열이 지원하는 저수준 연산으로 동일한 결과를 얻을 수 있습니다. 요소별 곱셈을 수행하고 결과 숫자의 합계를 계산합니다.

In [ ]:
(kanto * weights).sum()

56.8

** 연산자는 크기가 같은 두 배열의 요소별 곱셈을 수행합니다. sum 메서드는 배열에 있는 숫자의 합을 계산합니다.

In [ ]:
arr1 = np.array([1, 2, 3])
arr2 = np.array([4, 5, 6])

In [ ]:
arr1 * arr2

array([ 4, 10, 18])

In [ ]:
arr2.sum()

15

## 김소현 멘토 질문 : Numpy와 선형대수

### Numpy 배열 사용의 이점¶
Numpy 배열은 숫자 데이터 작업을 위해 Python 목록에 비해 다음과 같은 이점을 제공합니다.

사용 용이성: crop_yield와 같은 사용자 정의 함수 및 루프를 사용하는 대신 (kanto * weights).sum()과 같은 작고 간결하며 직관적인 수학 표현식을 작성할 수 있습니다.
성능: Numpy 연산 및 함수는 C++에서 내부적으로 구현되어 런타임에 해석되는 Python 문 및 루프를 사용하는 것보다 훨씬 빠릅니다.
다음은 각각 백만 개의 요소가 있는 두 벡터에서 Python 루프와 Numpy 배열을 사용하여 수행된 내적을 비교한 것입니다.

In [ ]:
# Python lists
arr1 = list(range(1000000))
arr2 = list(range(1000000, 2000000))

# Numpy arrays
arr1_np = np.array(arr1)
arr2_np = np.array(arr2)

In [ ]:
%%time
result = 0
for x1, x2 in zip(arr1, arr2):
    result += x1*x2
result

CPU times: user 151 ms, sys: 1.35 ms, total: 153 ms
Wall time: 152 ms


833332333333500000

In [ ]:
%%time
np.dot(arr1_np, arr2_np)

CPU times: user 1.96 ms, sys: 751 µs, total: 2.71 ms
Wall time: 1.6 ms


833332333333500000

보시다시피 np.dot를 사용하는 것은 for 루프를 사용하는 것보다 100배 빠릅니다. <br> 따라서 Numpy는 수만 또는 수백만 개의 데이터 포인트가 있는 매우 큰 데이터 세트로 작업할 때 특히 유용합니다.

계속하기 전에 작업을 저장합시다.

### 다차원 Numpy 배열
이제 한 단계 더 나아가 단일 2차원 Numpy 배열을 사용하여 모든 지역의 기후 데이터를 나타낼 수 있습니다.

In [ ]:
climate_data = np.array([[73, 67, 43],
                         [91, 88, 64],
                         [87, 134, 58],
                         [102, 43, 37],
                         [69, 96, 70]])

In [ ]:
climate_data

array([[ 73,  67,  43],
       [ 91,  88,  64],
       [ 87, 134,  58],
       [102,  43,  37],
       [ 69,  96,  70]])

고등학교에서 선형 대수학 수업을 들었다면 위의 2차원 배열을 5행 3열의 행렬로 인식할 수 있습니다. 각 행은 한 지역을 나타내고 열은 각각 기온, 강우량, 습도를 나타냅니다.

Numpy 배열은 여러 차원과 각 차원을 따라 다른 길이를 가질 수 있습니다. 배열의 .shape 속성을 사용하여 각 차원의 길이를 검사할 수 있습니다.
<img src="https://fgnt.github.io/python_crashkurs_doc/_images/numpy_array_t.png" width="420">

In [ ]:
# 2D array (matrix)
climate_data.shape

(5, 3)

In [ ]:
weights

array([0.3, 0.2, 0.5])

In [ ]:
# 1D array (vector)
weights.shape

(3,)

In [ ]:
# 3D array 
arr3 = np.array([
    [[11, 12, 13], 
     [13, 14, 15]], 
    [[15, 16, 17], 
     [17, 18, 19.5]]])

In [ ]:
arr3.shape

(2, 2, 3)

numpy 배열의 모든 요소는 동일한 데이터 유형을 갖습니다. .dtype 속성을 사용하여 배열의 데이터 유형을 확인할 수 있습니다.

In [ ]:
weights.dtype

dtype('float64')

In [ ]:
climate_data.dtype

dtype('int64')

배열에 단일 부동 소수점 숫자라도 포함되어 있으면 다른 모든 요소도 부동 소수점으로 변환됩니다.

In [ ]:
arr3.dtype

dtype('float64')

이제 `climate_data`(5x3 행렬)와 `weights`(길이 3의 벡터) 간의 단일 행렬 곱을 사용하여 모든 지역에서 예상되는 사과 수확량을 계산할 수 있습니다. <br>
시각적으로 보이는 것은 다음과 같습니다.
<img src="https://i.imgur.com/LJ2WKSI.png" width="240">

이 재생 목록의 처음 3-4개 비디오를 시청하여 행렬 및 행렬 곱셈에 대해 배울 수 있습니다. https://www.youtube.com/watch?v=xyAuNHPsq-g&list=PLFD0EB975BA0CC1E0&index=1 .

행렬 곱셈을 수행하기 위해 `np.matmul` 함수 또는 `@` 연산자를 사용할 수 있습니다.

In [ ]:
np.matmul(climate_data, weights)

array([56.8, 76.9, 81.9, 57.7, 74.9])

In [ ]:
climate_data @ weights

array([56.8, 76.9, 81.9, 57.7, 74.9])

### CSV 데이터 파일 작업
Numpy는 또한 파일에서 읽고 쓰는 도우미 기능을 제공합니다. 다음 형식으로 10,000개의 기후 측정값(온도, 강우량 및 습도)이 포함된 Climate.txt 파일을 다운로드해 보겠습니다.
```
온도, 강우량, 습도
25.00,76.00,99.00
39.00,65.00,70.00
59.00,45.00,77.00
84.00,63.00,38.00
66.00,50.00,52.00
41.00,94.00,77.00
91.00,57.00,96.00
49.00,96.00,99.00
67.00,20.00,28.00
...
```
이러한 데이터 저장 형식을 쉼표로 구분된 값 또는 CSV라고 합니다.

>CSV: 쉼표로 구분된 값(CSV) 파일은 쉼표를 사용하여 값을 구분하는 구분된 텍스트 파일입니다. 파일의 각 행은 데이터 레코드입니다. 각 레코드는 쉼표로 구분된 하나 이상의 필드로 구성됩니다. CSV 파일은 일반적으로 표 형식 데이터(숫자 및 텍스트)를 일반 텍스트로 저장하며, 이 경우 각 행에는 동일한 수의 필드가 있습니다. (위키피디아)

이 파일을 numpy 배열로 읽으려면 genfromtxt 함수를 사용할 수 있습니다.

In [ ]:

climate_data = np.genfromtxt(root_path + '/climate.txt', delimiter=',', skip_header=1)

In [ ]:
climate_data

array([[25., 76., 99.],
       [39., 65., 70.],
       [59., 45., 77.],
       ...,
       [99., 62., 58.],
       [70., 71., 91.],
       [92., 39., 76.]])

In [ ]:
climate_data.shape

(10000, 3)

이제 @ 연산자를 사용하여 행렬 곱을 수행하여 주어진 가중치 세트를 사용하여 전체 데이터 세트에 대한 사과 수확량을 예측할 수 있습니다.

In [ ]:
weights = np.array([0.3, 0.2, 0.5])

In [ ]:
yields = climate_data @ weights

In [ ]:
yields

array([72.2, 59.7, 65.2, ..., 71.1, 80.7, 73.4])

In [ ]:
yields.shape

(10000,)

np.concatenate 함수를 사용하여 네 번째 열로 climate_data에 수확량을 추가해 보겠습니다.

In [ ]:
climate_results = np.concatenate((climate_data, yields.reshape(10000, 1)), axis=1)

In [ ]:
climate_results

array([[25. , 76. , 99. , 72.2],
       [39. , 65. , 70. , 59.7],
       [59. , 45. , 77. , 65.2],
       ...,
       [99. , 62. , 58. , 71.1],
       [70. , 71. , 91. , 80.7],
       [92. , 39. , 76. , 73.4]])

여기에 몇 가지 미묘함이 있습니다.

새 열을 추가하고 싶기 때문에 axis=1 인수를 np.concatenate에 전달합니다. axis 인수는 연결의 차원을 지정합니다.

배열은 연결에 사용되는 차원을 제외하고 동일한 수의 차원과 동일한 길이를 가져야 합니다. np.reshape 함수를 사용하여 수확량의 모양을 (10000,)에서 (10000,1)로 변경합니다.

다음은 axis=1을 따라 np.concatenate에 대한 시각적 설명입니다(axis=0의 결과를 추측할 수 있습니까?):
<br>
<img src="https://www.w3resource.com/w3r_images/python-numpy-image-exercise-58.png" width="300">
<br>
Numpy 함수가 하는 일을 이해하는 가장 좋은 방법은 실험하고 문서를 읽고 인수 및 반환 값에 대해 배우는 것입니다. 

아래 셀을 사용하여 np.concatenate 및 np.reshape를 실험합니다.

위 계산의 최종 결과를 np.savetxt 함수를 사용하여 파일에 다시 작성해 보겠습니다.

In [ ]:
climate_results

array([[25. , 76. , 99. , 72.2],
       [39. , 65. , 70. , 59.7],
       [59. , 45. , 77. , 65.2],
       ...,
       [99. , 62. , 58. , 71.1],
       [70. , 71. , 91. , 80.7],
       [92. , 39. , 76. , 73.4]])

In [ ]:
np.savetxt(root_path+'/climate_results.txt', 
           climate_results, 
           fmt='%.2f', 
           delimiter=',',
           header='temperature,rainfall,humidity,yield_apples', 
           comments='')

결과는 CSV 형식으로 Climate_results.txt 파일에 다시 기록됩니다.
```
온도,강우량,습도,yield_apples
25.00,76.00,99.00,72.20
39.00,65.00,70.00,59.70
59.00,45.00,77.00,65.20
84.00,63.00,38.00,56.80
...
```

Numpy는 배열에서 작업을 수행하기 위한 수백 가지 기능을 제공합니다. 다음은 일반적으로 사용되는 몇 가지 기능입니다.

수학: np.sum, np.exp, np.round, 산술 연산자
배열 조작: np.reshape, np.stack, np.concatenate, np.split
선형 대수학: np.matmul, np.dot, np.transpose, np.eigvals
통계: np.mean, np.median, np.std, np.max
>필요한 기능을 찾는 방법은 무엇입니까? 특정 작업 또는 사용 사례에 적합한 기능을 찾는 가장 쉬운 방법은 웹 검색을 수행하는 것입니다. 예를 들어 "numpy 배열을 결합하는 방법"을 검색하면 배열 연결에 대한 이 자습서로 이동합니다.

여기에서 배열 함수의 전체 목록을 찾을 수 있습니다: https://numpy.org/doc/stable/reference/routines.html

### 산술 연산,  broadcasting 및 비교
Numpy 배열은 +, -, * 등과 같은 산술 연산자를 지원합니다. 단일 숫자(스칼라라고도 함) 또는 같은 모양의 다른 배열로 산술 연산을 수행할 수 있습니다. 연산자를 사용하면 다차원 배열로 수학 표현식을 쉽게 작성할 수 있습니다.

In [ ]:
arr2 = np.array([[1, 2, 3, 4], 
                 [5, 6, 7, 8], 
                 [9, 1, 2, 3]])

In [ ]:
arr3 = np.array([[11, 12, 13, 14], 
                 [15, 16, 17, 18], 
                 [19, 11, 12, 13]])

In [ ]:
# Adding a scalar
arr2 + 3

array([[ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12,  4,  5,  6]])

In [ ]:
# Element-wise subtraction
arr3 - arr2

array([[10, 10, 10, 10],
       [10, 10, 10, 10],
       [10, 10, 10, 10]])

In [ ]:
# Division by scalar
arr2 / 2

array([[0.5, 1. , 1.5, 2. ],
       [2.5, 3. , 3.5, 4. ],
       [4.5, 0.5, 1. , 1.5]])

In [ ]:
# Element-wise multiplication
arr2 * arr3

array([[ 11,  24,  39,  56],
       [ 75,  96, 119, 144],
       [171,  11,  24,  39]])

In [ ]:
# Modulus with scalar
arr2 % 4

array([[1, 2, 3, 0],
       [1, 2, 3, 0],
       [1, 1, 2, 3]])

### 어레이 브로드캐스팅
Numpy 배열은 또한 브로드캐스팅을 지원하여 차원 수는 다르지만 모양이 호환되는 두 배열 간의 산술 연산을 허용합니다. 어떻게 작동하는지 보기 위해 예제를 살펴보겠습니다.

In [ ]:
arr2 = np.array([[1, 2, 3, 4], 
                 [5, 6, 7, 8], 
                 [9, 1, 2, 3]])

In [ ]:
arr2.shape

(3, 4)

In [ ]:
arr4 = np.array([4, 5, 6, 7])

In [ ]:
arr4.shape

(4,)

In [ ]:
arr2 + arr4

array([[ 5,  7,  9, 11],
       [ 9, 11, 13, 15],
       [13,  6,  8, 10]])

arr2 + arr4 표현식이 평가될 때 arr2의 모양(3, 4)과 일치하도록 arr4(모양 (4,)를 가짐)가 세 번 복제됩니다. Numpy는 실제로 더 작은 차원 배열의 3개 사본을 생성하지 않고 복제를 수행하므로 성능이 향상되고 더 적은 메모리를 사용합니다.
<br>
<img src="https://jakevdp.github.io/PythonDataScienceHandbook/figures/02.05-broadcasting.png" width="360">
<br>
브로드캐스팅은 어레이 중 하나가 다른 어레이의 모양과 일치하도록 복제될 수 있는 경우에만 작동합니다.

In [ ]:
arr5 = np.array([7, 8])

In [ ]:
arr5.shape

(2,)

In [ ]:
arr2 + arr5

ValueError: operands could not be broadcast together with shapes (3,4) (2,) 

위의 예에서 rr5가 3번 복제되어도 arr2의 모양과 일치하지 않습니다. 따라서 rr2 + rr5는 성공적으로 평가될 수 없습니다. 여기에서 브로드캐스트에 대해 자세히 알아보세요. https://numpy.org/doc/stable/user/basics.broadcasting.html .

### 어레이 비교
Numpy 배열은 ==, !=, > 등과 같은 비교 연산도 지원합니다. 결과는 부울 배열입니다.

In [ ]:
arr1 = np.array([[1, 2, 3], [3, 4, 5]])
arr2 = np.array([[2, 2, 3], [1, 2, 5]])

In [ ]:
arr1 == arr2

array([[False,  True,  True],
       [False, False,  True]])

In [ ]:
arr1 != arr2

array([[ True, False, False],
       [ True,  True, False]])

In [ ]:
arr1 >= arr2

array([[False,  True,  True],
       [ True,  True,  True]])

In [ ]:
arr1 < arr2

array([[ True, False, False],
       [False, False, False]])

배열 비교는 sum 메서드를 사용하여 두 배열의 동일한 요소 수를 계산하는 데 자주 사용됩니다. 부울이 산술 연산에 사용될 때 True는 1로 평가되고 False는 0으로 평가된다는 점을 기억하십시오.

In [ ]:
(arr1 == arr2).sum()

3

### 배열 인덱싱 및 슬라이싱
Numpy는 `[]` 를 사용하여 Python의 목록 인덱싱 표기법을 직관적인 방식으로 다차원으로 확장합니다. Numpy 배열에서 특정 요소 또는 하위 배열(슬라이스라고도 함)을 선택하기 위해 쉼표로 구분된 인덱스 또는 범위 목록을 제공할 수 있습니다.

In [ ]:
arr3 = np.array([
    [[11, 12, 13, 14], 
     [13, 14, 15, 19]], 
    
    [[15, 16, 17, 21], 
     [63, 92, 36, 18]], 
    
    [[98, 32, 81, 23],      
     [17, 18, 19.5, 43]]])

In [ ]:
arr3.shape

(3, 2, 4)

In [ ]:
# Single element
arr3[1, 1, 2]

36.0

In [ ]:
# Subarray using ranges
arr3[1:, 0:1, :2]

array([[[15., 16.]],

       [[98., 32.]]])

In [ ]:
# Mixing indices and ranges
arr3[1:, 1, 3]

array([18., 43.])

In [ ]:
# Mixing indices and ranges
arr3[1:, 1, :3]

array([[63. , 92. , 36. ],
       [17. , 18. , 19.5]])

In [ ]:
# Using fewer indices
arr3[1]

array([[15., 16., 17., 21.],
       [63., 92., 36., 18.]])

In [ ]:
# Using fewer indices
arr3[:2, 1]

array([[13., 14., 15., 19.],
       [63., 92., 36., 18.]])

In [ ]:
# Using too many indices
arr3[1,3,2,1]

IndexError: too many indices for array: array is 3-dimensional, but 4 were indexed

표기법과 그 결과가 처음에는 혼란스러워 보일 수 있으므로 시간을 내어 실험하고 익숙해집니다. 아래 셀을 사용하여 다양한 인덱스 및 범위 조합을 사용하여 배열 인덱싱 및 슬라이싱의 몇 가지 예를 시도해 보십시오. 다음은 시각적으로 보여지는 몇 가지 예입니다.
<br>
<img src="https://scipy-lectures.org/_images/numpy_indexing.png" width="360">

### Numpy 배열을 만드는 다른 방법
Numpy는 또한 고정 또는 임의 값으로 원하는 모양의 배열을 생성하는 몇 가지 편리한 기능을 제공합니다. 공식 문서를 확인하거나 도움말 기능을 사용하여 자세히 알아보세요.

In [ ]:
# All zeros
np.zeros((3, 2))

array([[0., 0.],
       [0., 0.],
       [0., 0.]])

In [ ]:
# All ones
np.ones([2, 2, 3])

array([[[1., 1., 1.],
        [1., 1., 1.]],

       [[1., 1., 1.],
        [1., 1., 1.]]])

In [ ]:
# Identity matrix
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [ ]:
# Random vector
np.random.rand(5)

array([0.8208279 , 0.93801216, 0.44954753, 0.17816933, 0.32049174])

In [ ]:
# Random matrix
np.random.randn(2, 3) # rand vs. randn - what's the difference?

array([[ 0.83775   ,  1.13851471, -0.79147694],
       [ 0.56320765,  1.00386056, -0.42502339]])

In [ ]:
# Fixed value
np.full([2, 3], 42)

array([[42, 42, 42],
       [42, 42, 42]])

In [ ]:
# Range with start, end and step
np.arange(10, 90, 3)

array([10, 13, 16, 19, 22, 25, 28, 31, 34, 37, 40, 43, 46, 49, 52, 55, 58,
       61, 64, 67, 70, 73, 76, 79, 82, 85, 88])

In [ ]:
# Equally spaced numbers in a range
np.linspace(3, 27, 9)

array([ 3.,  6.,  9., 12., 15., 18., 21., 24., 27.])

# Exercises

1. 벡터란 무엇입니까?
1. Python 목록을 사용하여 벡터를 어떻게 표현합니까? 예를 들어 주십시오.
1. 두 벡터의 내적은 무엇입니까?
1. 두 벡터의 내적을 계산하는 함수를 작성하십시오.
1. Numpy 배열이란 무엇입니까?
1. Numpy 배열은 어떻게 만듭니까? 예를 들어 주십시오.
1. Numpy 배열의 유형은 무엇입니까?
1. Numpy 배열의 요소에 어떻게 액세스합니까?
1. Numpy를 사용하여 두 벡터의 내적을 어떻게 계산합니까?
1. 크기가 다른 두 벡터의 내적을 계산하려고 하면 어떻게 됩니까?
1. 두 Numpy 배열의 요소별 곱을 어떻게 계산합니까?
1. Numpy 배열의 모든 요소의 합을 어떻게 계산합니까?
1. 숫자 데이터 작업을 위해 Python 목록보다 Numpy 배열을 사용하면 어떤 이점이 있습니까?
1. Numpy 배열 작업이 Python 함수 및 루프에 비해 성능이 더 나은 이유는 무엇입니까?
1. 예제를 사용하여 Numpy 배열 작업과 Python 루프 간의 성능 차이를 보여줍니다.
1. 다차원 Numpy 배열이란 무엇입니까?
1. 2, 3, 4차원의 Numpy 배열 생성을 보여줍니다.
1. Numpy 배열에서 각 차원의 차원 수와 길이를 어떻게 검사합니까?
1. Numpy 배열의 요소가 다른 데이터 유형을 가질 수 있습니까?
1. Numpy 배열 요소의 데이터 유형을 어떻게 확인합니까?
1. Numpy 배열의 데이터 유형은 무엇입니까?
1. 행렬과 2D Numpy 배열의 차이점은 무엇입니까?
1. Numpy를 사용하여 행렬 곱셈을 어떻게 수행합니까?
1. Numpy에서 `@` 연산자는 무엇에 사용됩니까?
1. CSV 파일 형식은 무엇입니까?
1. Numpy를 사용하여 CSV 파일에서 데이터를 어떻게 읽습니까?
1. 두 개의 Numpy 배열을 어떻게 연결합니까?
1. `np.concatenate` 의 axis 인수의 목적은 무엇입니까?
1. 두 개의 Numpy 배열이 연결에 호환되는 경우는 언제입니까?
1. 연결할 수 있는 두 개의 Numpy 배열의 예를 제공하십시오.
1. 연결할 수 없는 두 개의 Numpy 배열의 예를 제공하십시오.
1. `np.reshape` 함수의 목적은 무엇입니까?
1. Numpy 배열을 "재구성"한다는 것은 무엇을 의미합니까?
1. CSV 파일에 numpy 배열을 어떻게 작성합니까?
1. 수학 연산을 수행하기 위한 Numpy 함수의 몇 가지 예를 제시하십시오.
1. 배열 조작을 수행하기 위한 Numpy 함수의 몇 가지 예를 제공하십시오.
1. 선형 대수를 수행하기 위한 Numpy 함수의 몇 가지 예를 제시하십시오.
1. 통계 연산을 수행하기 위한 Numpy 함수의 몇 가지 예를 제시하십시오.
1. 특정 작업이나 사용 사례에 적합한 Numpy 기능을 어떻게 찾습니까?
1. 모든 Numpy 배열 함수 및 작업 목록은 어디에서 볼 수 있습니까?
1. Numpy 배열에서 지원하는 산술 연산자는 무엇입니까? 예를 들어 설명합니다.
1. 배열 방송이란 무엇입니까? 어떻게 유용합니까? 예를 들어 설명합니다.
1. 브로드캐스팅과 호환되는 어레이의 몇 가지 예를 들어 주시겠습니까?
1. 브로드캐스팅과 호환되지 않는 배열의 예를 몇 가지 제시하십시오.
1. Numpy 배열에서 지원하는 비교 연산자는 무엇입니까? 예를 들어 설명합니다.
1. Numpy 배열에서 특정 하위 배열 또는 슬라이스에 어떻게 액세스합니까?
1. 몇 가지 예를 들어 다차원 Numpy 배열에서 배열 인덱싱 및 슬라이싱을 설명합니다.
1. 모든 0을 포함하는 주어진 모양으로 Numpy 배열을 어떻게 생성합니까?
1. 모든 것을 포함하는 주어진 모양으로 Numpy 배열을 어떻게 생성합니까?
1. 주어진 모양의 단위 행렬을 어떻게 생성합니까?
1. 주어진 길이의 임의 벡터를 어떻게 생성합니까?
1. 각 요소에 대해 고정 값을 사용하여 주어진 모양으로 Numpy 배열을 어떻게 생성합니까?
1. 무작위로 초기화된 요소를 포함하는 주어진 모양으로 Numpy 배열을 어떻게 생성합니까?
1. `np.random.rand` 와 `np.random.randn` 의 차이점은 무엇입니까? 예를 들어 설명합니다.
1. `np.arange` 와 `np.linspace` 의 차이점은 무엇입니까? 예를 들어 설명합니다.